packages/storage/test/unit/reference.exp.test.ts

```diff
@@ -0,0 +1,556 @@
+import { DEFAULT_HOST } from '../../src/implementation/constants';
+import { FirebaseAuthInternalName } from '@firebase/auth-interop-types';
+import { Provider } from '@firebase/component';
+
+/* eslint-disable @typescript-eslint/no-floating-promises */
+function makeFakeService(
+  app: FirebaseApp,
+  authProvider: Provider<FirebaseAuthInternalName>,
+  sendHook: SendHook
+): StorageService {
+  return new StorageService(app, authProvider, testShared.makePool(sendHook));
+}
+
+function makeStorage(url: string): Reference {
+  const service = new StorageService(
+    {} as FirebaseApp,
+    testShared.emptyAuthProvider,
+    testShared.makePool(null)
+  );
+  return new Reference(service, url);
+}
+
+describe('Firebase Storage > Reference', () => {
+  const root = makeStorage('gs://test-bucket/');
+  const child = makeStorage('gs://test-bucket/hello');
+  describe('Path constructor', () => {
+    it('root', () => {
+      assert.equal(root.toString(), 'gs://test-bucket/');
+    });
+    it('keeps characters after ? on a gs:// string', () => {
+      const s = makeStorage('gs://test-bucket/this/ismyobject?hello');
+      assert.equal(s.toString(), 'gs://test-bucket/this/ismyobject?hello');
+    });
+    it("doesn't URL-decode on a gs:// string", () => {
+      const s = makeStorage('gs://test-bucket/%3F');
+      assert.equal(s.toString(), 'gs://test-bucket/%3F');
+    });
+    it('ignores URL params and fragments on an http URL', () => {
+      const s = makeStorage(
+        `http://${DEFAULT_HOST}/v0/b/test-bucket/o/my/object.txt` +
+          '?ignoreme#please'
+      );
+      assert.equal(s.toString(), 'gs://test-bucket/my/object.txt');
+    });
+    it('URL-decodes and ignores fragment on an http URL', () => {
+      const s = makeStorage(
+        `http://${DEFAULT_HOST}/v0/b/test-bucket/o/%3F?ignore`
+      );
+      assert.equal(s.toString(), 'gs://test-bucket/?');
+    });
+
+    it('ignores URL params and fragments on an https URL', () => {
+      const s = makeStorage(
+        `https://${DEFAULT_HOST}/v0/b/test-bucket/o/my/object.txt` +
+          '?ignoreme#please'
+      );
+      assert.equal(s.toString(), 'gs://test-bucket/my/object.txt');
+    });
+
+    it('URL-decodes and ignores fragment on an https URL', () => {
+      const s = makeStorage(
+        `https://${DEFAULT_HOST}/v0/b/test-bucket/o/%3F?ignore`
+      );
+      assert.equal(s.toString(), 'gs://test-bucket/?');
+    });
+  });
+
+  describe('toString', () => {
+    it("Doesn't add trailing slash", () => {
+      const s = makeStorage('gs://test-bucket/foo');
+      assert.equal(s.toString(), 'gs://test-bucket/foo');
+    });
+    it('Strips trailing slash', () => {
+      const s = makeStorage('gs://test-bucket/foo/');
+      assert.equal(s.toString(), 'gs://test-bucket/foo');
+    });
+  });
+
+  describe('parentReference', () => {
+    it('Returns null at root', () => {
+      assert.isNull(parentReference(root));
```

Comment:
FWIW, we moved `parent` back to its existing location for Firestore, which would match `root`.